### PR TITLE
Sort list of tasks and display them by earlier deadline first.

### PR DESCRIPTION
**Pull request by Kang-Quan**
Adding of tasks will now be O(log(n)) where n is size of taskList instead of O(1) previously to maintain the sorted list.

Similarly resetData() function in TaskList will now be O(nlogn) instead of O(n).

### CODE REVIEW

File: src/main/java/seedu/address/model/ReadOnlyTaskList.java [hunk 1]
```diff
@@ -10,7 +10,7 @@ public interface ReadOnlyTaskList {
 
     /**
      * Returns an unmodifiable view of the task list.
-     * This list will not contain any duplicate persons.
+     * This list will not contain any duplicate tasks.
```
Comment (hauchongtang) [inline src/main/java/seedu/address/model/ReadOnlyTaskList.java:13]: Good find !

File: src/main/java/seedu/address/model/TaskList.java [hunk 2]
```diff
@@ -87,6 +92,48 @@ public void deleteTask(Index index) {
...
+
+            if (currentDate.isAfter(givenDate) || currentDate.isEqual(givenDate)) {
+                high = mid - 1;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        return low;
+    }
```
Comment (hauchongtang) [inline src/main/java/seedu/address/model/TaskList.java:135]: Good addition but remember to test it thoroughly for potential edge cases.

File: src/test/java/seedu/address/model/task/TaskListTest.java [hunk 3]
```diff
@@ -4,8 +4,12 @@
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static seedu.address.testutil.Assert.assertThrows;
+import static seedu.address.testutil.TypicalTasks.TASK_DEADLINE_AFTER_CHICKEN;
+import static seedu.address.testutil.TypicalTasks.TASK_DEADLINE_BEFORE_CHICKEN;
+import static seedu.address.testutil.TypicalTasks.TASK_DEADLINE_SAME_CHICKEN;
```
Comment (hauchongtang) [inline src/test/java/seedu/address/model/task/TaskListTest.java:9]: Good abstraction

File: src/test/java/seedu/address/model/task/TaskListTest.java [hunk 4]
```diff
@@ -107,4 +112,33 @@ public void deleteTask_validIndex_success() {
...
+        int expectedResults = 1;
+        assertEquals(results, expectedResults);
+    }
+
+    @Test
+    public void findIndexOfTask_emptyTaskList_success() {
+        int results = taskList.findIndexOfTask(TASK_DEADLINE_AFTER_CHICKEN);
+        int expectedResults = 0;
+        assertEquals(results, expectedResults);
+    }
```
Comment (hauchongtang) [inline src/test/java/seedu/address/model/task/TaskListTest.java:142]: Comprehensive test cases, good to see that test cases check for empty task list as well.

File: src/test/java/seedu/address/testutil/TypicalTasks.java [hunk 5]
```diff
@@ -41,4 +59,19 @@ public static TaskList getTypicalTaskList() {
...
+    /**
+     * Returns a TaskList used to test the order of deadline.
+     */
+    public static TaskList getTimedTaskList() {
+        TaskList taskList = new TaskList();
+        getTimedTasks().forEach(task -> {
+            taskList.addTask(task);
+        });
+        return taskList;
+    }
```
Comment (hauchongtang) [inline src/test/java/seedu/address/testutil/TypicalTasks.java:76]: Good separation of concerns.

**Review by hauchongtang — APPROVED** [review]
LGTM ! Include your feature in the UG/DG if needed.
